### DISCUSSION TITLE
Fatal Error when calling GetSourceAsync()

### DISCUSSION THREAD

**Discussion by huazai76** [cefsharp/CefSharp#3838] [General]
- **What version of the product are you using?**
94.4.2+g6a963ca+chromium-94.0.4606.61

- **What architecture x86 or x64?**
    x64   - note: I also tried x86, and a similar error occured
	
- **What version of .Net?**
    .Net 4.6
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WinForms
    
- **What steps will reproduce the problem?**
    When executing following codes (for scrawling webpages) on a particular URL, an error occurs:
```c#            
CefBrowser.Load("https://www.epri.com/research/products/3002012302");
Application.DoEvents();
Thread.Sleep(8000);

try
{
  Task<string> t = CefBrowser.GetSourceAsync();
  t.Wait();   
  var h = t.Result;
}
catch (Exception ex)
{
    MessageBox.Show(ex.Message);
}
```

- **What is the expected output? What do you see instead?**
Error Messages are copied and translated into English (marked with **) below:

System.AccessViolationException
  HResult=0x80004003
  Message= ** try to read or write protected memory,.... 
  Source=mscorlib
  StackTrace:
   at System.String.wcslen(Char* ptr)
   at System.String.CtorCharPtr(Char* ptr)
   at CefSharp.Internals.CefStringVisitorAdapter.Visit(CefStringVisitorAdapter* , CefStringBase<CefStringTraitsUTF16>* string) in C:\projects\cefsharp\packages\cef.sdk.93.1.14\CEF\include\internal\cef_string_wrappers.h:line 447

  ** this exception is initiated at the following calling stack :
	string.wcslen(char*)
	string.CtorCharPtr(char*)
    CefSharp::Internals::CefStringVisitorAdapter::Visit(CefStringBase<CefStringTraitsUTF16>&) (in cef_string_wrappers.h)

**Comment by amaitland** [comment]
Converting to discussion as it's not clear this is a bug. You shouldn't rely on Thread.Sleep for loading of a website.

You can use LoadUrlAsync to wait for the website to load.
http://cefsharp.github.io/api/94.4.x/html/M_CefSharp_IWebBrowser_LoadUrlAsync.htm

GetSourceAsync is an asynchronous operation, Continue With or await should be used rather than blocking.

You can add a breakpoint at https://github.com/cefsharp/CefSharp/blob/cefsharp/94/CefSharp.Core.Runtime/Internals/CefStringVisitorAdapter.h#L33 debug the underlying exception.

**Reply by amaitland** [reply]
See https://github.com/cefsharp/CefSharp/blob/a3504aa5ad2de33082bc614675b7638dca7c8dbb/CefSharp.Test/OffScreen/OffScreenBrowserBasicFacts.cs#L451 for an example.

You can try running the test with the `Url` you've provided, I suspect it will pass.

**Reply by huazai76** [reply]
Much thanks.   Below is further infomation:

I am a beginner in multi-thread programming, and up to now I don't know how to correctly use the "await" statement.  I do understand that using "sleep(8000)" is not a good way, but I have tried "await" (just as your example code) a few days before, and cannot pass the compilor syntax checking, so I have to use "sleep".   And, since I am sure the page content is loaded completely before executing GetSource, I dont expect the "await" method will not fail in this senario.
And, the reason I believe this is an underlying C++ Bug, is that I have use this code to scrawl many identical WebPage in this website, and all other pages succeed, up to now, noly this one failed.   So I think possibly some special contents /characters in this particular page  triggered the bug.

The URL list of the web pages I scrawlled is in this index page: 
https://www.epri.com/research/sectors/technology/results

**Reply by amaitland** [reply]
> but I have tried "await" (just as your example code) a few days before, and cannot pass the compilor syntax checking, so I have to use "sleep"

If you require generally programming help then asking on https://stackoverflow.com/questions/tagged/cefsharp is the best place. Add tags for the `c#` language.

Using `await` or `ContinueWith` are the only supported options. Using `Thread.Sleep` doesn't guarantee the page has loaded, regardless of how long you sleep for. The website may have failed to load in which case you cannot get the source. If you use `LoadUrlAsync` you can easily deal with errors. It's also possible the render process might have crashed. Quite simply you cannot use `Thread.Sleep` as you have reliably.

> And, the reason I believe this is an underlying C++ Bug,

Possibly though not necessarily a bug in `CefSharp`. It's more likely a bug in the `Chromium Embedded Framework(CEF)`.

If you can provide a failing `xUnit` test then issue #3837 can be reopened. As it stands the example you've provided isn't sufficient for me to investigate this further.

**Reply by huazai76** [reply]
My actual code use a loop to check the content in HTML to ensure that the whole page are downloaded prior to GETSOURCE, and the page have be rendered and shown on screen a few seconds before executing GETSOURCE,  so the SLEEP cannot be a problem here (even though it is not an eligible code).  

I am an amateur, and have no experience in UNIT TEST.  Maybe later I will have a try.

**Reply by ukandrewc** [reply]
Put your code in the DocumentComplete event, that tells you the page is loaded.

**Reply by amaitland** [reply]
You can use http://cefsharp.github.io/api/94.4.x/html/E_CefSharp_IWebBrowser_LoadingStateChanged.htm or http://cefsharp.github.io/api/94.4.x/html/E_CefSharp_IWebBrowser_FrameLoadEnd.htm depending on your requirements. 

> My actual code use a loop to check the content in HTML to ensure that the whole page are downloaded prior to GETSOURCE, 

Have you tested the code you provided above and confirmed that the error reproduces in isolation? Sounds like your code is far more complex than the example provided. 

See https://stackoverflow.com/help/minimal-reproducible-example for pointers on providing a reproducible example. 

How exactly are you checking the html content before calling GetSourceAsync?

**Reply by huazai76** [reply]
My actual code is a little bit more complex, so after encountering the exception, I used a simpler code (as provided in the above bug report), for making sure which statement causes the exception and communicating this issue clearly later.   And this simple code reproduced the same error, even when I have changed the SLEEP time much longer.  These code is wrapped in a separate function and called by the MAIN FORM thread(hence isolated?).  I have used breakpoint to trace it by executing the code step by step. 

ps:  Thanks for your several good programming advices.  I believe those are correct and elegant ways.  However, since I am just an ameteur and C# beginner, right now, I am not able to write down muliti-thread and async code in this good way.  That is why I still stick with the ugly "SLEEP" solution,  and hope it will work (actually it work for all the other identical webpages).
---
- How exactly are you checking the html content before calling GetSourceAsync?
My code for checking:
                do
                {
                    Thread.Sleep(100);
                    Task<string> taskGetSrc = CefBrowser.GetSourceAsync();
                    taskGetSrc.Wait();     
                    html = taskGetSrc.Result;
                }while (!(html.Contains("<div class=\"details\">"));  
----
To  ukandrewc:
I dont believe the error is caused by completeness of page loading.   And I think even the page is not loaded completely, the program will not break down in this way, i.e.  the exception cannot be catched in my C# code (I have tried to add an TRY-CATCH statement), and the EXE program quit just silently, without any error message dialog shown (when the program is started in Windows instead started from Visual Studio IDE).

**Reply by amaitland** [reply]
> I used a simpler code (as provided in the above bug report),

Please fork https://github.com/cefsharp/CefSharp.MinimalExample and push your changes to GitHub. Only make the minimum of changes required to reproduce the problem,  squash all changes into a single commit. I will then have a look.

Ultimately this is unlikely a bug in CefSharp, potentially the error handling can be improved, I'll need an example that reproduces the problem to know for sure.

**Reply by huazai76** [reply]
> > I used a simpler code (as provided in the above bug report),
> 
> Please fork https://github.com/cefsharp/CefSharp.MinimalExample and push your changes to GitHub. Only make the minimum of changes required to reproduce the problem, squash all changes into a single commit. I will then have a look.
> 
> Ultimately this is unlikely a bug in CefSharp, potentially the error handling can be improved, I'll need an example that reproduces the problem to know for sure.

I have try the minimal winform example, I put a button on it, and execute same code I have provided, and it worked.
So, then I check the browser configuration differents between two apps.  And finally find the "CachePath " settiing cause this error.
When I remove the  CachePath = cacheDir  statement in the following code, app function ok.  
So, there may be a bug in the memory from /to disk seriazation code.  

            CefSettings setting = new CefSettings()
            {
                Locale = "zh-CN",                    
                AcceptLanguageList = "zh-CN",
                CachePath = cacheDir,      //--- TRIGGERING BUG  to be removed , 
                LocalesDirPath = appPath + "locales\\",
            };
            Cef.Initialize(setting);

--- note :     The cacheDir  is set in a sub-dir named "cache" in my app path, so should be OK for itself.
-- note:  I tried to remove all other setting and keep the CachePath setting, the ERROR still occur.,   So  CachePath is the ONLY problematic statement.

**Reply by huazai76** [reply]
Still don't work for another URL. Same exception come out.   After testing this case,  it seem the content of page ( string returned by  GetSource ) keep changing before end of loading and script execution, and it changes many times, in some cases, it first contain the keyword, then not ,and then in the end contain the keyword.  So the way using if (Contain( "someKeywords"))   to determine the page is completely loaded does not work.  That means in some particular pages, I get the wrong source from intermidiate page content.

Thanks to amaitland  and ukandrewc.   Possibly you are right,  the program will only work completely in the way not using the SLEEP.

**Reply by ukandrewc** [reply]
Just tried your URL and got the source in FrameLoadEnd, every time.

**Reply by amaitland** [reply]
> I have try the minimal winform example, I put a button on it, and execute same code I have provided, and it worked.

The minimum example has a CachePath set by default. Please focus on reproducing the problem with the minimum example. I need a fork that reliably reproduces the problem before I can debug this and see if we can handle the exception gracefully.

**Reply by huazai76** [reply]
> The minimum example has a CachePath set by default. Please focus on reproducing the problem with the minimum example. I need a fork that reliably reproduces the problem before I can debug this and see if we can handle the exception gracefully.

I works hours and cannot exactly locate the exceptiontiggering code.  Up to now, by removing code sections each time and test,  I think possibly three factors together can triggering the exception: 
(1)  cachdir is set;
(2)  A FormResize methord defined, and some controls properties is set and got, such as width, clientwidth, 
(3)  CefBrowser.Dock = DockStyle.None,  -- this shoud have a connection with factor 2.

After a long time editing, I have messed up my codes, maybe later I can provide the exact example.

> GetSourceAsync is an asynchronous operation, Continue With or await should be used rather than blocking.

Besides, the  **GetSourceAsync()**  cannot not be used with **await**  as the sample code provide you above,  because, C# require the await followed by a method modified by " **Async**" ,,,  however, this method is not defined like that as shown by the method name. The compilor  raise an error for  await GetSourceAsync().  Could you pls provide me some more details about how to use GetSourceAsync  with await?

**Reply by amaitland** [reply]
> Besides, the **GetSourceAsync()** cannot not be used with **await** as the sample code provide you above, because, C# require the await followed by a method modified by " **Async**" ,,

I'm not sure what reference you are quoting here. You can absolutely use `GetSourceAsync()` with `await` as it returns `Task<string>`. For an `event handler` you simply need to add the `async` keyword to the method declaration. 

```c#
private async void BtnStart_Click(object sender, EventArgs e)
{
	await browser.LoadUrlAsync("https://www.epri.com/research/products/1013030");
	string html = "";
	do
	{
		html = await browser.GetSourceAsync();
	}
	while (!html.Contains("<h2 class=\"header\">Keywords</h2>"));

	MessageBox.Show(html);
}
```

**Comment by huazai76** [comment]
I tried but fail to use the WinformMinimum to trigger the exception.  By removing codes of my app, I have got a very simple code that will trigger the exeption.  But I cannot upload a attachment by dragging & dropping here.   So I copy all the code below.
---code deleted.   more concise code provided below.

**Reply by amaitland** [reply]
Please add your form to the minimal example and make any of the modifications nessicary.

**Reply by huazai76** [reply]
Sorry, I have no experience in version control or forking..     I have delete some more unnecessay codes and make a most concise version, and re-test to make sure that the three conditions altogether will trigger the exception.   If removing anyone, no exception..   
I have make comments in the code, showing the details of the controls on the form.  I believe it is easy task for you as to re-construct the project.  

---

```c#
using System;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.IO;
using CefSharp;
using CefSharp.WinForms;

namespace EPRIProductsCrawler
{
    public partial class FrmMain : Form
    {
        private readonly ChromiumWebBrowser browser;  

        public FrmMain()
        {
            InitializeComponent();

            Cef.EnableHighDPISupport();
            CefSettings setting = new CefSettings()
            {
                CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache"),
            };
            Cef.Initialize(setting, performDependencyCheck: false);

            browser = new ChromiumWebBrowser("about:version")
            {
                Dock = DockStyle.None,
            };
            this.Controls.Add(browser);
        }
        
        private void FrmMain_Resize(object sender, EventArgs e)
        {
            // The controls in and the arragement of FrmMain Form are: 
            //upper  is a label (LblToolbar) used as toolbar, and a textbox for show URL;  a Panel (PnlMain) docked left; 
            // and a CEFBrowser whose location is calculated by FrmMain_Resize method
            if (browser != null)
            {
                const int MIN_SIZE = 50;
                this.SuspendLayout();
                browser.Left =  PnlMain.Left + PnlMain.Width ;
                browser.Top = LblToolbar.Top + LblToolbar.Height;
                browser.Width = Math.Max(ClientSize.Width - browser.Left, MIN_SIZE);
                browser.Height = Math.Max(ClientSize.Height - browser.Top - StatusStrip.Height, MIN_SIZE);
                TxtURL.Width = Math.Max(ClientSize.Width - TxtURL.Left - 10, MIN_SIZE);
                this.ResumeLayout(false);
                this.PerformLayout();
            }
        }

        private void FrmMain_FormClosing(object sender, FormClosingEventArgs e)
        {
            browser.Dispose();
            System.Environment.Exit(0);  
        }
        
        private void BtnStart_Click(object sender, EventArgs e)
        {
            //This will triggering an  System.AccessViolationException  that is not handled and cannot be catched, and thus is fatal.
            // URLs triggering exception include at least：  1013030   3002012302    3002014760  1011829  1021228  
            // If remove brower's CachePath setting or Dock.none setting, or remove FrmMain_Resize(), no exception will occur

            browser.Load("https://www.epri.com/research/products/1013030");  
            string h ="";
            do
            {
                System.Threading.Thread.Sleep(1000);
                Application.DoEvents();
                try
                {
                    Task<string> t = browser.GetSourceAsync();
                    t.Wait();     
                    h = t.Result;
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message);
                }
            } while (!h.Contains("<h2 class=\"header\">Keywords</h2>"));
            MessageBox.Show(h);   // for making sure no exception occur before
        }     
    }
}
```

**Reply by amaitland** [reply]
> Sorry, I have no experience in version control or forking

https://docs.github.com/en/get-started/quickstart/fork-a-repo

> The controls in and the arragement of FrmMain Form are: 
        //upper  is a label (LblToolbar) used as toolbar, and a textbox for show URL;  a Panel (PnlMain) docked left; 
        // and a CEFBrowser whose location is calculated by FrmMain_Resize 

This is the reason I ask for a fork of the minimum example. Trying to reconstruct an example takes time and is open to interpretation. I'll test without this code as I'm not going to try implemting this portion. 

If this portion is relevant then you'll need to provide the designer code as well. (No zip files as I won't download them from an unknown source).

**Comment by huazai76** [comment]
further info:
In all the 3000 identical URLs crawled,   three webpage trigger the exception (and always trigger).   They may have some common attibutes 
URLs  are :   1013030   3002012302    3002014760   (plus the prefix: https://www.epri.com/research/products/ )

**Comment by huazai76** [comment]
More info:
In the three factors, pls focus on   **Dock = DockStyle.None**   if you debug later.   Because later I find other two URLs (1011829 1021228) that cause exception even when I remove the Form_Resize method and the CachePath setting ..    And after Dock = DockStyle.None is also removed, no exception occur for these two URL. 
But there is a weird thing, these two URL only still cause exception after removing CachePath in my actual code ( with a loop to crawl many pages),  but do not cause exception in the test code after removing CachePath.   This is just FYI, and I will not further investigate the reason for this weird issue, it is too complicated for me.   If you need further info, I can try my best.     I will just remove all these three code sections in my apps to aviod crash and hope no more other exceptions come.

**Reply by amaitland** [reply]
> In the three factors, pls focus on **Dock = DockStyle.None** if you debug later. Because later I find other two URLs (1011829 1021228) that cause exception even when I remove the Form_Resize method and the CachePath setting .. And after Dock = DockStyle.None is also removed, no exception occur for these two URL.

I don't believe these are actually related to the exception you are seeing, it's possible they somehow impact the timing in someway. It's very unlikely they are directly related. 

I'd suggest you test with the `async` code I provided above.

> Where appropriate, we now utilize the default UTF string encoding format and
shared memory to reduce copies and conversions for the cross-process
transfer of arbitrary-length strings. For example, CefFrame::GetSource/GetText
now involves zero UTF conversions and zero copies in the browser process for
the CefString delivered to CefStringVisitor::Visit().

As per https://github.com/chromiumembedded/cef/commit/ebee84755ed14e71388b343231d3a419f1c5c1fd

Looking at the `CEF` history this change in behaviour to `CefStringVisitor::Visit()` seems the most likely cause.

Based on the exception my best guess is that the shared memory section used for the string somehow becomes invalid. Maybe the render process crashes and the memory is unavailable? Without an example I have no way to test a possible solution, the best I can do is take a guess.

**Reply by amaitland** [reply]
Commit https://github.com/cefsharp/CefSharp/commit/7722881772c462d83b275d568f4f341fc89eb26d changes the way string are handled in an attempt to get a copy of the string rather than using a ptr to workaround the `upstream` issue.

You can test out a nightly build for [94.4.20-CI4254](https://www.myget.org/feed/cefsharp/package/nuget/CefSharp.WinForms/94.4.20-CI4254), Instructions for installing the package via the `Nuget Console` are included on the page.

As I have no way of testing this all that I know is that it passes the existing unit tests.

**Reply by huazai76** [reply]
I got the new CEF package.   And use the my minimum code to test all the konwn problematic URL.   And it is OK now.  No exception.!! 
So the three factors should only be contributing and indirect factors, not the root cause.     You might have eliminate the root cause.  

Excepetion URLs tested：  1013030   3002012302    3002014760  1011829  1021228    1013030

**Comment by pkudrel** [comment]
I can confirm that the problem exists. 
But my application is complicated and it is not easy to isolate the problem. I will try to reproduce the issue using `WinformsMinimum` example.

**Reply by amaitland** [reply]
Test with the build linked in https://github.com/cefsharp/CefSharp/discussions/3838#discussioncomment-1422707

**Reply by pkudrel** [reply]
Thank You - I'll do it.
But I want to give an example that generates the problem.

**Comment by amaitland** [comment]
> Commit [7722881](https://github.com/cefsharp/CefSharp/commit/7722881772c462d83b275d568f4f341fc89eb26d) changes the way string are handled in an attempt to get a copy of the string rather than using a ptr to workaround the `upstream` issue.

Changes were included in the `94.4.50` release.
